sklearn/preprocessing/imputation.py

```diff
@@ -374,3 +528,454 @@ def transform(self, X):
+            equal to max_neighbors and if calling method is fit_transform()
+            then its value must be equal to max_neighbors + 1.
+        """
+
+        check_is_fitted(self, 'statistics_')
+        force_all_finite = False if self.missing_values in ["NaN",
+                                                            np.nan] else True
+        X = check_array(X, accept_sparse=False, dtype=FLOAT_DTYPES,
+                        force_all_finite=force_all_finite, copy=self.copy)
+        # Check for +/- inf
+        if np.any(np.isinf(X)):
+            raise ValueError("+/- inf values are not allowed in data to be "
+                             "transformed.")
+
+        # Get fitted data and ensure correct dimension
+        fitted_X = self._fitted_neighbors._fit_X
+        if X.shape[1] != fitted_X.shape[1]:
+            raise ValueError("Incompatible dimension between the fitted "
+                             "dataset and the one to be transformed.")
+        mask = _get_mask(X, self.missing_values)
+        n_rows_X, n_cols_X = X.shape
+        row_total_missing = mask.sum(axis=1)
+        if not np.any(row_total_missing):
+            return X
+
+        # Check for excessive missingness in rows
+        bad_rows = row_total_missing > (mask.shape[1] * self.row_max_missing)
+        if np.any(bad_rows):
+            warnings.warn(
+                "There are rows with more than {0}% missing values. The "
+                "missing features in these rows are imputed with column means."
+                    .format(self.row_max_missing * 100))
+            X_bad = X[bad_rows, :]
+            X = X[~bad_rows, :]
+            mask = mask[~bad_rows]
+            row_total_missing = mask.sum(axis=1)
+        row_has_missing = row_total_missing.astype(np.bool)
+
+        if np.any(row_has_missing):
+            if self.use_complete:
+                # Initializations
+                # Mask for fitted_X
+                mask_fx = _get_mask(fitted_X, np.nan)
+
+                # Locate unique patterns
+                patterns, row_pat_idx = _unique(
+                    mask, return_inverse=True, axis=0)
+
+                # Get row idx for receivers (missing)
+                receiver_row_missing_index, _ = np.where(mask)
+
+                # For every pattern, index receivers and potential donors
+                for p in range(len(patterns)):
+                    if not np.any(patterns[p]):
+                        continue
+                    # receivers are those with pattern 'p'
+                    row_has_missing_pat = (row_pat_idx == p)
+
+                    # Donors have features missing in receivers available
+                    # The bitwise-AND captures if something is missing in both
+                    donor_row_idx = ~np.any(patterns[p] & mask_fx, axis=1)
+
+                    # Change donor set to ones not missing relevant features
+                    self._fitted_neighbors._fit_X = fitted_X[donor_row_idx]
```

Comment:
This is not a good idea. Just construct a new NearestNeighbors. Or don't use NearestNeighbors at all.

Reply:
I don't like how it hacks the private interface. But I suppose it avoids re-validating the input data

Reply:
I see. But as you indicated it saves compute time, which already would have increased compared to the previous implementation due to a loopy operation over each missing pattern. Also if the concern is over inadvertently changing the original data, I could work with a .copy() instead, although that of course increases memory footprint (and i also doubt this is the source of your concern). Another option is to add an explicit setter method for fitted_X in Nearest Neighbors. 
Anyway, I will move ahead once you guys decide on the appropriate way forward.

Reply:
Shrug. Leave it as is for now.

Reply:
Gotcha. And does the rest of the algorithm, broadly speaking, look okay to you? Also @amueller, if you could please take a look and let me know if you are okay with the current implementation, I would much appreciate it.

Reply:
Minus the required cleanups and streamlining of things, of course.

Reply:
By the way, do we keep both Nearest Neighbors and Nearest Available Neighbors setting latter as default? FWIW, I personally think that this is a good idea given that the "official" implementation is still the former. But please advise on how you guys want to proceed with the implementation here.